File: src/CapturePlus/Core/SelectionTracker.cs
using System;

namespace CapturePlus.Core;

public sealed class SelectionTracker
{
    private double _startX, _startY;

    public NormRect Rect { get; private set; }
    public bool IsDragging { get; private set; }
    public bool IsLocked { get; private set; }

    public void Begin(double x, double y)
    {
        if (IsLocked) return;
        _startX = x;
        _startY = y;
        Rect = new NormRect(x, y, 0, 0);
        IsDragging = true;
    }

    public void Update(double x, double y)
    {
        if (!IsDragging || IsLocked) return;
        Rect = SelectionNormalizer.Normalize(_startX, _startY, x - _startX, y - _startY);
    }

    public bool End()
    {
        if (!IsDragging || IsLocked) return false;
        IsDragging = false;
        if (SelectionNormalizer.IsValid(Rect.X, Rect.Y, Rect.Width, Rect.Height))
        {
            IsLocked = true;
            return true;
        }
        Reset();
        return false;
    }

    public void Reset()
    {
        IsDragging = false;
        IsLocked = false;
        Rect = default;
    }
}

File: src/SelectionTracker.cpp
#include "SelectionTracker.h"

namespace selnorm {

NormRect normalize(double x, double y, double w, double h)
{
    if (w < 0) { x += w; w = -w; }
    if (h < 0) { y += h; h = -h; }
    return NormRect(x, y, w, h);
}

bool isValid(double w, double h) { return w >= 8 && h >= 8; }

}

void SelectionTracker::begin(double x, double y)
{
    if (locked_) return;
    startX_ = x; startY_ = y;
    rect_ = NormRect(x, y, 0, 0);
    dragging_ = true;
}

void SelectionTracker::update(double x, double y)
{
    if (!dragging_ || locked_) return;
    rect_ = selnorm::normalize(startX_, startY_, x - startX_, y - startY_);
}

bool SelectionTracker::end()
{
    if (!dragging_ || locked_) return false;
    dragging_ = false;
    if (selnorm::isValid(rect_.w, rect_.h))
    {
        locked_ = true;
        return true;
    }
    reset();
    return false;
}

void SelectionTracker::reset()
{
    dragging_ = false;
    locked_   = false;
    rect_     = NormRect();
}
